File: examples/api/privatools.py
#!/usr/bin/env python3
"""PrivaTools async PDF starter. Python 3.10+, standard library only."""
from __future__ import annotations

import argparse
import contextlib
import email.utils
import hashlib
import http.client
import json
import os
from pathlib import Path
import random
import re
import socket
import sys
import tempfile
import threading
import time
import urllib.error
import urllib.parse
import urllib.request
import uuid

DEFAULT_BASE = "https://api.privatools.me/api/v1"
MAX_INPUT = 50 * 1024 * 1024
MAX_RESULT = 100 * 1024 * 1024
OPERATIONS = ("merge", "compress", "pdf-to-text")
LEVELS = ("light", "recommended", "extreme", "email", "print", "archive", "web")
ID = re.compile(r"[A-Za-z0-9_-]{1,128}\Z")


class StarterError(Exception):
    pass


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None  # Never forward the key to a redirected origin.


class TimedResponse(http.client.HTTPResponse):
    def close(self):
        timer = getattr(self, "starter_timer", None)
        if timer:
            timer.cancel()
        super().close()


class TimedConnection:
    """A socket inactivity timeout alone cannot stop a slow-drip response."""
    response_class = TimedResponse

    def __init__(self, *args, deadline, **kwargs):
        self.starter_deadline = deadline
        self.starter_timer = None
        super().__init__(*args, **kwargs)

    def connect(self):
        super().connect()
        sock = self.sock
        remaining = self.starter_deadline - time.monotonic()
        if remaining <= 0:
            sock.close()
            raise TimeoutError("Connection deadline reached")

        def stop_socket():
            with contextlib.suppress(OSError):
                sock.shutdown(socket.SHUT_RDWR)

        self.starter_timer = threading.Timer(remaining, stop_socket)
        self.starter_timer.daemon = True
        self.starter_timer.start()

    def getresponse(self):
        response = super().getresponse()
        response.starter_timer = self.starter_timer
        return response


class TimedHTTPConnection(TimedConnection, http.client.HTTPConnection):
    pass


class TimedHTTPSConnection(TimedConnection, http.client.HTTPSConnection):
    pass


class TimedHTTPHandler(urllib.request.HTTPHandler):
    def __init__(self, deadline):
        super().__init__()
        self.deadline = deadline

    def http_open(self, request):
        return self.do_open(TimedHTTPConnection, request, deadline=self.deadline)


class TimedHTTPSHandler(urllib.request.HTTPSHandler):
    def __init__(self, deadline):
        super().__init__()
        self.deadline = deadline

    def https_open(self, request):
        return self.do_open(TimedHTTPSConnection, request, deadline=self.deadline, context=self._context)


def base_url(value):
    parsed = urllib.parse.urlsplit(value)
    if (parsed.scheme not in ("https", "http") or not parsed.hostname
            or parsed.username or parsed.password or parsed.query or parsed.fragment
            or (parsed.scheme == "http" and parsed.hostname not in ("localhost", "127.0.0.1", "::1"))):
        raise StarterError("Use an HTTPS base URL without credentials, query, or fragment (HTTP is allowed only on loopback).")
    return value.rstrip("/")


def job_id(value, key=None):
    if not isinstance(value, str) or not ID.fullmatch(value) or key and key in value:
        raise StarterError("The API returned an invalid job ID.")
    return value


def retry_delay(headers, fallback):
    raw = headers.get("Retry-After", "")
    try:
        delay = float(raw)
        if delay >= 0 and delay < float("inf"):
            return delay
    except ValueError:
        pass
    try:
        date = email.utils.parsedate_to_datetime(raw)
        return max(0, date.timestamp() - time.time())
    except (ValueError, TypeError, OverflowError):
        return fallback


class Client:
    def __init__(self, base, key, deadline_seconds=1200):
        self.base = base_url(base)
        if not key or any(ord(c) < 33 or ord(c) > 126 for c in key):
            raise StarterError("Set PRIVATOOLS_API_KEY to your API key.")
        self.key = key
        self.deadline = time.monotonic() + deadline_seconds

    def remaining(self):
        remaining = self.deadline - time.monotonic()
        if remaining <= 0:
            raise StarterError("Deadline reached. Your saved job state can be resumed; the result was not deleted.")
        return remaining

    def pause(self, seconds):
        if seconds >= self.remaining():
            raise StarterError("Retry-After exceeds the remaining deadline. Resume later using your saved state.")
        time.sleep(max(0.05, seconds))

    def request(self, method, route, body=None, headers=None):
        # All routes are built locally; status_url/result.url are never followed.
        if not re.fullmatch(r"jobs(?:/[A-Za-z0-9_-]{1,128}(?:/result)?)?", route):
            raise StarterError("Invalid local API route.")
        for attempt in range(5):
            request = urllib.request.Request(self.base + "/" + route, data=body, method=method,
                                             headers={"X-API-Key": self.key, **(headers or {})})
            try:
                timeout = min(30, self.remaining())
                deadline = time.monotonic() + timeout
                opener = urllib.request.build_opener(NoRedirect(), TimedHTTPHandler(deadline), TimedHTTPSHandler(deadline))
                return opener.open(request, timeout=timeout)
            except urllib.error.HTTPError as error:
                status = error.code
                delay = retry_delay(error.headers, 2 ** attempt + random.random() / 4)
                # Do not print response bodies, which could contain reflected secrets.
                request_id = error.headers.get("X-Request-ID", "")
                error.close()
                if status not in (429, 502, 503, 504) or attempt == 4:
                    suffix = f" Request ID: {request_id}" if ID.fullmatch(request_id) and self.key not in request_id else ""
                    raise StarterError(f"API HTTP {status}.{suffix} Saved state is available for recovery.") from None
            except (urllib.error.URLError, TimeoutError, OSError, http.client.HTTPException):
                if attempt == 4:
                    raise StarterError("Network retry budget exhausted. Resume with the saved state.") from None
                delay = 2 ** attempt + random.random() / 4
            self.pause(delay)
        raise AssertionError("unreachable")

    def json(self, method, route, body=None, headers=None):
        with self.request(method, route, body, headers) as response:
            chunks = []
            count = 0
            while True:
                self.remaining()
                chunk = response.read1(64 * 1024)
                if not chunk:
                    break
                count += len(chunk)
                if count > 1024 * 1024:
                    raise StarterError("Unexpectedly large API metadata response.")
                chunks.append(chunk)
            data = b"".join(chunks)
            if len(data) > 1024 * 1024:
                raise StarterError("Unexpectedly large API metadata response.")
            try:
                value = json.loads(data)
            except (ValueError, UnicodeError):
                raise StarterError("Invalid JSON metadata response. Resume using the saved state.") from None
            if not isinstance(value, dict):
                raise StarterError("Invalid JSON metadata object. Resume using the saved state.")
            return value, response.headers


def save_state(path, state, new=False):
    # State contains IDs, options, and hashes; never the API key or document paths.
    encoded = (json.dumps(state, indent=2) + "\n").encode()
    if new:
        fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
        with os.fdopen(fd, "wb") as target:
            target.write(encoded)
            target.flush()
            os.fsync(target.fileno())
        return
    parent = Path(path).resolve().parent
    fd, tmp = tempfile.mkstemp(prefix=".priva-state-", dir=parent)
    try:
        with os.fdopen(fd, "wb") as target:
            target.write(encoded)
            target.flush()
            os.fsync(target.fileno())
        os.replace(tmp, path)
    finally:
        with contextlib.suppress(FileNotFoundError):
            os.unlink(tmp)


def load_state(path):
    try:
        state = json.loads(Path(path).read_text())
        if (state["version"] != 1 or state["operation"] not in OPERATIONS
                or not ID.fullmatch(state["idempotencyKey"])):
            raise ValueError()
        base_url(state["baseUrl"])
        if state.get("jobId"):
            job_id(state["jobId"])
        return state
    except (KeyError, ValueError, TypeError):
        raise StarterError("Invalid state file.") from None


def prepare_upload(operation, filenames, level):
    minimum, maximum = (2, 10) if operation == "merge" else (1, 1)
    if not minimum <= len(filenames) <= maximum:
        raise StarterError(f"{operation} requires {minimum}–{maximum} PDF inputs.")
    if sum(Path(name).stat().st_size for name in filenames) > MAX_INPUT:
        raise StarterError("Total PDF input exceeds 50 MiB.")
    blobs = [Path(name).read_bytes() for name in filenames]
    if sum(map(len, blobs)) > MAX_INPUT or any(not blob.startswith(b"%PDF-") for blob in blobs):
        raise StarterError("Inputs must be PDF files totaling at most 50 MiB.")
    options = {"level": level} if operation == "compress" else {}
    hashes = [{"sha256": hashlib.sha256(blob).hexdigest(), "bytes": len(blob)} for blob in blobs]
    boundary = "priva-" + uuid.uuid4().hex
    chunks = []
    for name, value in (("operation", operation), ("options", json.dumps(options, sort_keys=True))):
        chunks.append(f'--{boundary}\r\nContent-Disposition: form-data; name="{name}"\r\n\r\n{value}\r\n'.encode())
    for index, blob in enumerate(blobs):
        chunks.extend([f'--{boundary}\r\nContent-Disposition: form-data; name="files"; filename="input-{index+1}.pdf"\r\nContent-Type: application/pdf\r\n\r\n'.encode(), blob, b"\r\n"])
    chunks.append(f"--{boundary}--\r\n".encode())
    return b"".join(chunks), "multipart/form-data; boundary=" + boundary, options, hashes


def wait_for_job(client, identifier):
    for _ in range(600):
        job, headers = client.json("GET", "jobs/" + job_id(identifier))
        if job.get("id") != identifier:
            raise StarterError("Job ID mismatch.")
        state = job.get("state")
        if state == "succeeded":
            return job
        if state in ("failed", "canceled", "expired"):
            code = (job.get("error") or {}).get("code", "")
            safe_code = " (" + code + ")" if isinstance(code, str) and re.fullmatch(r"[a-z0-9_]{1,100}", code) and client.key not in code else ""
            raise StarterError(f"Job {state}{safe_code}. No result was deleted by this client.")
        if state not in ("queued", "running"):
            raise StarterError("Unknown job state.")
        client.pause(max(0.1, retry_delay(headers, 2)))
    raise StarterError("Polling budget exhausted. Resume using your saved state.")


def download(client, job, output):
    result = job.get("result") or {}
    expected = result.get("bytes")
    if type(expected) is not int or not 0 < expected <= MAX_RESULT:
        raise StarterError("Invalid result size metadata.")
    output = Path(output)
    if output.exists():
        raise StarterError("Output already exists. Choose another output path; the server result is retained.")
    fd, tmp = tempfile.mkstemp(prefix=".priva-download-", dir=output.resolve().parent)
    try:
        count = 0
        with os.fdopen(fd, "wb") as target, client.request("GET", "jobs/" + job_id(job["id"]) + "/result") as response:
            while True:
                client.remaining()
                chunk = response.read1(256 * 1024)
                if not chunk:
                    break
                count += len(chunk)
                if count > expected:
                    raise StarterError("Result exceeds expected size. The server result is retained.")
                target.write(chunk)
            if count != expected:
                raise StarterError("Result download is incomplete. Resume to download again.")
            target.flush()
            os.fsync(target.fileno())
        # Link fails if another process created the destination during download.
        os.link(tmp, output)
    finally:
        with contextlib.suppress(FileNotFoundError):
            os.unlink(tmp)


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("command", choices=(*OPERATIONS, "resume", "delete"))
    parser.add_argument("files", nargs="*")
    parser.add_argument("--output", "-o", help="New local output path (.json for pdf-to-text)")
    parser.add_argument("--state", help="State JSON path; reuse it only for the same logical request")
    parser.add_argument("--base-url", help="Explicit API base; HTTPS or loopback HTTP")
    parser.add_argument("--level", choices=LEVELS, default="recommended")
    parser.add_argument("--retain", action="store_true", help="Keep server result until expiry instead of deleting after save")
    parser.add_argument("--deadline-seconds", type=float, default=1200)
    args = parser.parse_args(argv)
    if not 0 < args.deadline_seconds <= 3600:
        raise StarterError("Deadline must be positive and at most 3600 seconds.")
    if args.command != "delete" and not args.output:
        raise StarterError("Provide --output (use a .json filename for pdf-to-text).")
    if args.command in ("resume", "delete") and (not args.state or args.files):
        raise StarterError("Resume/delete requires --state and no file arguments.")
    state_path = args.state or f"privatools-job-{uuid.uuid4().hex}.json"
    existing = Path(state_path).exists()
    if args.command in ("resume", "delete") and not existing:
        raise StarterError("State file does not exist.")
    state = load_state(state_path) if existing else None
    base = base_url(args.base_url or (state["baseUrl"] if state else DEFAULT_BASE))
    if state and base != state["baseUrl"]:
        raise StarterError("The selected API base differs from the saved request. Use a new state file for a different server.")
    client = Client(base, os.environ.get("PRIVATOOLS_API_KEY", ""), args.deadline_seconds)
    if args.command in OPERATIONS:
        body, content_type, options, hashes = prepare_upload(args.command, args.files, args.level)
        if state:
            if (state["operation"], state["options"], state["inputs"]) != (args.command, options, hashes):
                raise StarterError("This state belongs to different input/options. Use a new state file for new work.")
        else:
            state = {"version": 1, "baseUrl": base, "operation": args.command, "options": options,
                     "inputs": hashes, "idempotencyKey": uuid.uuid4().hex, "jobId": None}
            save_state(state_path, state, new=True)
        print(f"State: {state_path}", flush=True)
        print(f"Idempotency key: {state['idempotencyKey']}", flush=True)
        if not state.get("jobId"):
            job, _ = client.json("POST", "jobs", body, {"Content-Type": content_type, "Idempotency-Key": state["idempotencyKey"]})
            state["jobId"] = job_id(job.get("id"), client.key)
            print(f"Job ID: {state['jobId']}", flush=True)
            save_state(state_path, state)
    if not state or not state.get("jobId"):
        raise StarterError("Submission was interrupted before saving a job ID. Repeat the original file command with this --state path.")
    job_id(state["jobId"], client.key)
    print(f"Job ID: {state['jobId']}", flush=True)
    if args.command == "delete":
        deleted, _ = client.json("DELETE", "jobs/" + state["jobId"])
        print("Cancellation requested; poll to confirm cleanup." if deleted.get("deletion_pending") else "Server result deleted.")
        return
    job = wait_for_job(client, state["jobId"])
    download(client, job, args.output)
    print(f"Saved: {args.output}", flush=True)
    if args.retain:
        print("Server result retained until its normal expiry (one hour after completion). Use delete to remove it sooner.")
    else:
        deleted, _ = client.json("DELETE", "jobs/" + state["jobId"])
        print("Deletion pending." if deleted.get("deletion_pending") else "Server result deleted after successful save.")


if __name__ == "__main__":
    try:
        main()
    except (StarterError, OSError, ValueError, http.client.HTTPException, KeyboardInterrupt) as error:
        # Local filesystem errors can include paths, but no headers or API key.
        message = str(error) if isinstance(error, StarterError) else "Local IO/response error or interruption. The server result was not automatically deleted; use saved state to recover."
        print("Error: " + message, file=sys.stderr)
        sys.exit(1)

File: frontend/public/api-starters/privatools.mjs
#!/usr/bin/env node
/** PrivaTools async PDF starter. Node.js 22+, no packages required. */
import { createHash, randomUUID } from 'node:crypto';
import { open, readFile, stat, rename, unlink, link } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { pathToFileURL } from 'node:url';
import { parseArgs } from 'node:util';

export const DEFAULT_BASE = 'https://api.privatools.me/api/v1';
const MAX_INPUT = 50 * 1024 * 1024;
const MAX_RESULT = 100 * 1024 * 1024;
const OPERATIONS = ['merge', 'compress', 'pdf-to-text'];
const LEVELS = ['light', 'recommended', 'extreme', 'email', 'print', 'archive', 'web'];
const ID = /^[A-Za-z0-9_-]{1,128}$/;
export class StarterError extends Error {}

export function baseUrl(value) {
  let url;
  try { url = new URL(value); } catch { throw new StarterError('Invalid API base URL.'); }
  if (!['https:', 'http:'].includes(url.protocol) || url.username || url.password || url.search || url.hash
      || (url.protocol === 'http:' && !['localhost', '127.0.0.1', '[::1]'].includes(url.hostname))) {
    throw new StarterError('Use HTTPS without credentials, query, or fragment (HTTP is allowed only on loopback).');
  }
  return value.replace(/\/+$/, '');
}

function jobId(value, key) {
  if (typeof value !== 'string' || !ID.test(value) || key && value.includes(key)) throw new StarterError('The API returned an invalid job ID.');
  return value;
}

export function retryDelay(headers, fallback) {
  const raw = headers.get('retry-after');
  if (raw !== null && raw.trim() !== '') {
    const seconds = Number(raw);
    if (Number.isFinite(seconds) && seconds >= 0) return seconds;
    const timestamp = Date.parse(raw);
    if (Number.isFinite(timestamp)) return Math.max(0, (timestamp - Date.now()) / 1000);
  }
  return fallback;
}

export class Client {
  constructor(base, key, deadlineSeconds = 1200) {
    this.base = baseUrl(base);
    if (!key || !/^[\x21-\x7e]+$/.test(key)) throw new StarterError('Set PRIVATOOLS_API_KEY to your API key.');
    this.key = key;
    this.deadline = performance.now() + deadlineSeconds * 1000;
  }
  remaining() {
    const remaining = (this.deadline - performance.now()) / 1000;
    if (remaining <= 0) throw new StarterError('Deadline reached. Resume using the saved state; the result was not deleted.');
    return remaining;
  }
  async pause(seconds) {
    if (seconds >= this.remaining()) throw new StarterError('Retry-After exceeds the remaining deadline. Resume later using your saved state.');
    await new Promise((done) => setTimeout(done, Math.max(50, seconds * 1000)));
  }
  async request(method, route, body, headers = {}) {
    // Ignore every URL in API responses. Build authenticated routes locally.
    if (!/^jobs(?:\/[A-Za-z0-9_-]{1,128}(?:\/result)?)?$/.test(route)) throw new StarterError('Invalid local API route.');
    for (let attempt = 0; attempt < 5; attempt++) {
      let response;
      let delay = 2 ** attempt + Math.random() / 4;
      const timeout = Math.max(1, Math.ceil(Math.min(30, this.remaining()) * 1000));
      try {
        response = await fetch(`${this.base}/${route}`, {
          method, body, headers: { 'X-API-Key': this.key, ...headers },
          redirect: 'manual', signal: AbortSignal.timeout(timeout),
        });
      } catch {
        if (attempt === 4) throw new StarterError('Network retry budget exhausted. Resume using the saved state.');
      }
      if (response) {
        if (response.ok) return response;
        delay = retryDelay(response.headers, delay);
        await response.body?.cancel();
        if (![429, 502, 503, 504].includes(response.status) || attempt === 4) {
          const requestId = response.headers.get('x-request-id') || '';
          const suffix = ID.test(requestId) && !requestId.includes(this.key) ? ` Request ID: ${requestId}` : '';
          throw new StarterError(`API HTTP ${response.status}.${suffix} Saved state is available for recovery.`);
        }
      }
      await this.pause(delay);
    }
    throw new StarterError('Retry budget exhausted.');
  }
  async json(method, route, body, headers) {
    const response = await this.request(method, route, body, headers);
    const chunks = [];
    let count = 0;
    for await (const chunk of response.body) {
      this.remaining();
      count += chunk.length;
      if (count > 1024 * 1024) throw new StarterError('Unexpectedly large API metadata response.');
      chunks.push(chunk);
    }
    let value;
    try { value = JSON.parse(Buffer.concat(chunks).toString('utf8')); }
    catch { throw new StarterError('Invalid JSON metadata response. Resume using the saved state.'); }
    return [value, response.headers];
  }
}

async function exists(path) {
  try { await stat(path); return true; } catch (error) { if (error.code === 'ENOENT') return false; throw error; }
}

export async function saveState(path, state, isNew = false) {
  // No API key, document filenames, or filesystem input paths are persisted.
  const destination = isNew ? path : resolve(dirname(path), `.priva-state-${randomUUID()}`);
  const handle = await open(destination, 'wx', 0o600);
  try {
    await handle.writeFile(JSON.stringify(state, null, 2) + '\n');
    await handle.sync();
  } finally { await handle.close(); }
  if (!isNew) {
    try { await rename(destination, path); }
    finally { await unlink(destination).catch((error) => { if (error.code !== 'ENOENT') throw error; }); }
  }
}

async function loadState(path) {
  let state;
  try { state = JSON.parse(await readFile(path, 'utf8')); }
  catch { throw new StarterError('Cannot read state JSON.'); }
  if (state?.version !== 1 || !OPERATIONS.includes(state.operation) || !ID.test(state.idempotencyKey)) throw new StarterError('Invalid state file.');
  baseUrl(state.baseUrl);
  if (state.jobId) jobId(state.jobId);
  return state;
}

async function prepareUpload(operation, filenames, level) {
  const [minimum, maximum] = operation === 'merge' ? [2, 10] : [1, 1];
  if (filenames.length < minimum || filenames.length > maximum) throw new StarterError(`${operation} requires ${minimum}–${maximum} PDF inputs.`);
  let size = 0;
  for (const name of filenames) size += (await stat(name)).size;
  if (size > MAX_INPUT) throw new StarterError('Total PDF input exceeds 50 MiB.');
  const blobs = await Promise.all(filenames.map((name) => readFile(name)));
  if (blobs.reduce((sum, blob) => sum + blob.length, 0) > MAX_INPUT || blobs.some((blob) => blob.subarray(0, 5).toString() !== '%PDF-')) {
    throw new StarterError('Inputs must be PDF files totaling at most 50 MiB.');
  }
  const options = operation === 'compress' ? { level } : {};
  const hashes = blobs.map((blob) => ({ sha256: createHash('sha256').update(blob).digest('hex'), bytes: blob.length }));
  const form = new FormData();
  form.append('operation', operation);
  form.append('options', JSON.stringify(options));
  blobs.forEach((blob, index) => form.append('files', new Blob([blob], { type: 'application/pdf' }), `input-${index + 1}.pdf`));
  return { form, options, hashes };
}

export async function waitForJob(client, identifier) {
  for (let count = 0; count < 600; count++) {
    const [job, headers] = await client.json('GET', `jobs/${jobId(identifier)}`);
    if (job.id !== identifier) throw new StarterError('Job ID mismatch.');
    if (job.state === 'succeeded') return job;
    if (['failed', 'canceled', 'expired'].includes(job.state)) {
      const code = job.error?.code;
      const suffix = typeof code === 'string' && /^[a-z0-9_]{1,100}$/.test(code) && !code.includes(client.key) ? ` (${code})` : '';
      throw new StarterError(`Job ${job.state}${suffix}. No result was deleted by this client.`);
    }
    if (!['queued', 'running'].includes(job.state)) throw new StarterError('Unknown job state.');
    await client.pause(Math.max(0.1, retryDelay(headers, 2)));
  }
  throw new StarterError('Polling budget exhausted. Resume using your saved state.');
}

export async function download(client, job, output) {
  const expected = job.result?.bytes;
  if (!Number.isInteger(expected) || expected <= 0 || expected > MAX_RESULT) throw new StarterError('Invalid result size metadata.');
  if (await exists(output)) throw new StarterError('Output already exists. Choose another output path; the server result is retained.');
  const temporary = resolve(dirname(output), `.priva-download-${randomUUID()}`);
  const handle = await open(temporary, 'wx', 0o600);
  try {
    const response = await client.request('GET', `jobs/${jobId(job.id)}/result`);
    let count = 0;
    for await (const chunk of response.body) {
      client.remaining();
      count += chunk.length;
      if (count > expected) throw new StarterError('Result exceeds expected size. The server result is retained.');
      await handle.writeFile(chunk);
    }
    if (count !== expected) throw new StarterError('Result download is incomplete. Resume to download again.');
    await handle.sync();
    // A hard link publishes complete bytes atomically and never replaces a file.
    await link(temporary, output);
  } finally {
    await handle.close();
    await unlink(temporary).catch((error) => { if (error.code !== 'ENOENT') throw error; });
  }
}

export async function main(args = process.argv.slice(2)) {
  const { values, positionals } = parseArgs({ args, allowPositionals: true, options: {
    output: { type: 'string', short: 'o' }, state: { type: 'string' },
    'base-url': { type: 'string' }, level: { type: 'string', default: 'recommended' },
    retain: { type: 'boolean', default: false }, 'deadline-seconds': { type: 'string', default: '1200' },
    help: { type: 'boolean', short: 'h' },
  } });
  if (values.help) {
    console.log('node privatools.mjs <merge|compress|pdf-to-text> FILE... --output FILE [--state JOB.json] [--retain]\nnode privatools.mjs resume --state JOB.json --output FILE [--retain]\nnode privatools.mjs delete --state JOB.json\nSet PRIVATOOLS_API_KEY. Options: --base-url URL --level recommended --deadline-seconds 1200');
    return;
  }
  const [command, ...files] = positionals;
  if (![...OPERATIONS, 'resume', 'delete'].includes(command)) throw new StarterError('Choose merge, compress, pdf-to-text, resume, or delete. Use --help.');
  if (!LEVELS.includes(values.level)) throw new StarterError('Invalid compression level.');
  const deadline = Number(values['deadline-seconds']);
  if (!Number.isFinite(deadline) || deadline <= 0 || deadline > 3600) throw new StarterError('Deadline must be positive and at most 3600 seconds.');
  if (command !== 'delete' && !values.output) throw new StarterError('Provide --output (use a .json filename for pdf-to-text).');
  if (['resume', 'delete'].includes(command) && (!values.state || files.length)) throw new StarterError('Resume/delete requires --state and no file arguments.');
  const statePath = values.state || `privatools-job-${randomUUID().replaceAll('-', '')}.json`;
  const existing = await exists(statePath);
  if (['resume', 'delete'].includes(command) && !existing) throw new StarterError('State file does not exist.');
  let state = existing ? await loadState(statePath) : null;
  const base = baseUrl(values['base-url'] || state?.baseUrl || DEFAULT_BASE);
  if (state && base !== state.baseUrl) throw new StarterError('The API base differs from the saved request. Use new state for a different server.');
  const client = new Client(base, process.env.PRIVATOOLS_API_KEY, deadline);
  if (OPERATIONS.includes(command)) {
    const { form, options, hashes } = await prepareUpload(command, files, values.level);
    if (state) {
      if (state.operation !== command || JSON.stringify(state.options) !== JSON.stringify(options) || JSON.stringify(state.inputs) !== JSON.stringify(hashes)) {
        throw new StarterError('This state belongs to different input/options. Use a new state file for new work.');
      }
    } else {
      state = { version: 1, baseUrl: base, operation: command, options, inputs: hashes,
        idempotencyKey: randomUUID().replaceAll('-', ''), jobId: null };
      await saveState(statePath, state, true);
    }
    console.log(`State: ${statePath}\nIdempotency key: ${state.idempotencyKey}`);
    if (!state.jobId) {
      const [job] = await client.json('POST', 'jobs', form, { 'Idempotency-Key': state.idempotencyKey });
      state.jobId = jobId(job.id, client.key);
      console.log(`Job ID: ${state.jobId}`);
      await saveState(statePath, state);
    }
  }
  if (!state?.jobId) throw new StarterError('Submission was interrupted before saving a job ID. Repeat the original file command with this --state path.');
  jobId(state.jobId, client.key);
  console.log(`Job ID: ${state.jobId}`);
  if (command === 'delete') {
    const [deleted] = await client.json('DELETE', `jobs/${state.jobId}`);
    console.log(deleted.deletion_pending ? 'Cancellation requested; poll to confirm cleanup.' : 'Server result deleted.');
    return;
  }
  const job = await waitForJob(client, state.jobId);
  await download(client, job, values.output);
  console.log(`Saved: ${values.output}`);
  if (values.retain) console.log('Server result retained until its normal expiry (one hour after completion). Use delete to remove it sooner.');
  else {
    const [deleted] = await client.json('DELETE', `jobs/${state.jobId}`);
    console.log(deleted.deletion_pending ? 'Deletion pending.' : 'Server result deleted after successful save.');
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(resolve(process.argv[1])).href) {
  main().catch((error) => {
    // Never log fetch errors, request objects, headers, or reflected response bodies.
    console.error(`Error: ${error instanceof StarterError ? error.message : 'Local IO/response error. The server result was not automatically deleted; use saved state to recover.'}`);
    process.exitCode = 1;
  });
}
